### PR TITLE
Auto-reload synonym analyzers on synonyms updates

### PR DESCRIPTION
**Pull request by mayya-sharipova**
Synonym Management API project

On changes of synonyms in a synonym set, auto-reload analyzers. 
Note that currently all updateable analyzers will be reloaded, even
those that are not relevant for a synonyms set being updated.

Responses from PUT and DELETE synonyms set are enhanced to 
contain reload analyzers details.

For a PUT synonyms set request, an example of response:
PUT _synonyms/set1
```js
{
    "result": "created",
    "reload_analyzers_details": {
        "_shards": {
            "total": 0,
            "successful": 0,
            "failed": 0
        },
        "reload_details": []
    }
}
```
or an update:

```js
{
    "result": "updated",
    "reload_analyzers_details": {
        "_shards": {
            "total": 2,
            "successful": 1,
            "failed": 0
        },
        "reload_details": [
            {
                "index": "index1",
                "reloaded_analyzers": [
                    "my_analyzer"
                ],
                "reloaded_node_ids": [
                    "vxKVyIRJQuO4DPj8Y70ijg"
                ]
            }
        ]
    }
}
```

For a DELETE synonyms set request, if the deleted synonyms set is in use, a response 
will contain reload failures, like below:
```js
{
    "acknowledged": true,
    "reload_analyzers_details": {
        "_shards": {
            "total": 2,
            "successful": 0,
            "failed": 1,
            "failures": [
                {
                    "shard": 0,
                    "index": "index1",
                    "status": "NOT_FOUND",
                    "reason": {
                        "type": "resource_not_found_exception",
                        "reason": "Synonym set [set1] not found"
                    }
                }
            ]
        },
        "reload_details": []
    }
}
```
And it is **left for a user to fix affected indices**.

### CODE REVIEW

File: modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/70_synonyms_from_index.yml [hunk 1]
```diff
@@ -66,6 +69,71 @@
...
+        body:
+          synonyms_set:
+            - synonyms: "hello, salute"
+            - synonyms: "ciao => goodbye"
+  - match: { result: "updated" }
+  - gte: { reload_analyzers_details._shards.total: 1 }
+  - match: { reload_analyzers_details.reload_details.0.index: "my_index" }
+  - match: { reload_analyzers_details.reload_details.0.reloaded_analyzers.0 : "my_analyzer" }
+
+  # Confirm that the index analyzers are reloaded
```
Comment (carlosdelest) [inline modules/analysis-common/src/yamlRestTest/resources/rest-api-spec/test/analysis-common/70_synonyms_from_index.yml:85]: I like this test ❤️

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml [hunk 2]
```diff
@@ -13,6 +13,8 @@
               id: "test-id"
 
   - match: { result: "created" }
+  - match: { reload_analyzers_details._shards.total: 0 }
```
Comment (carlosdelest) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml:16]: I'm not sure we should include reload analyzer details on create. WDYT?

File: server/src/main/java/org/elasticsearch/action/synonyms/DeleteSynonymsAction.java [hunk 3]
```diff
@@ -71,4 +78,54 @@ public int hashCode() {
...
+            this.reloadAnalyzersResponse = reloadAnalyzersResponse;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+            builder.startObject();
+            builder.field(ACKNOWLEDGED_KEY, acknowledgedResponse.isAcknowledged());
+            builder.field("reload_analyzers_details");
+            reloadAnalyzersResponse.toXContent(builder, params);
+            builder.endObject();
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/action/synonyms/DeleteSynonymsAction.java:107]:
```suggestion
            builder.startObject();
            {
                        builder.field(ACKNOWLEDGED_KEY, acknowledgedResponse.isAcknowledged());
                        builder.field("reload_analyzers_details");
                        reloadAnalyzersResponse.toXContent(builder, params);
            }
            builder.endObject();
```

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java [hunk 4]
```diff
@@ -243,8 +246,20 @@ public void putSynonymsSet(String resourceName, SynonymRule[] synonymsSet, Actio
...
                 .execute(deleteByQueryResponseListener.delegateFailure((bulkResponseListener, bulkResponse) -> {
                     if (bulkResponse.hasFailures() == false) {
-                        UpdateSynonymsResult result = created ? UpdateSynonymsResult.CREATED : UpdateSynonymsResult.UPDATED;
-                        bulkResponseListener.onResponse(result);
+                        UpdateSynonymsResultStatus updateSynonymsResultStatus = created
+                            ? UpdateSynonymsResultStatus.CREATED
+                            : UpdateSynonymsResultStatus.UPDATED;
+
+                        // auto-reload all reloadable analyzers (currently only those that use updateable synonym or keyword_marker filters)
+                        // TODO: reload only those analyzers that use this synonymsSet
```
Comment (carlosdelest) [inline server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:254]: What are your thoughts for reloading just the analyzers for the specific syonyms set? Do it in a follow up PR? Or do we need some additional infra for retrieving the analyzers that use it?

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml (hunk 2 above)
Comment (mayya-sharipova) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml:16]: I thought about this, but than I thought it is better for response to look the same and not different depending on conditions. What do you think?

File: server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java (hunk 4 above)
Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java:254]: Right, we can do this in a follow-up. It is not trivial how to do this efficiently.

**Review by carlosdelest — APPROVED** [review]

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml (hunk 2 above)
Comment (mayya-sharipova) [reply rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/synonyms/10_synonyms_put.yml:16]: Synchronized with @carlosdelest offline, and we decided to keep the response as is for 8.9, and may revisit for GA.

**Pull request merged**